hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java

```diff
@@ -153,9 +194,194 @@ public BackgroundTaskResult call() {
+                      .getBlocksForKeyDelete(deletedKey);
+                  if (blocksForKeyDelete != null) {
+                    blockGroupList.addAll(blocksForKeyDelete);
+                  }
+                  delCount++;
+                } else {
+                  newRepeatedOmKeyInfo.addOmKeyInfo(keyInfo);
+                }
+              }
+
+              if (newRepeatedOmKeyInfo.getOmKeyInfoList().size() > 0 &&
+                  newRepeatedOmKeyInfo.getOmKeyInfoList().size() !=
+                  repeatedOmKeyInfo.getOmKeyInfoList().size()) {
+                keysToModify.put(deletedKey, newRepeatedOmKeyInfo);
+              }
+
+              if (newRepeatedOmKeyInfo.getOmKeyInfoList().size() !=
+                  repeatedOmKeyInfo.getOmKeyInfoList().size()) {
+                keysToPurge.addAll(blockGroupList);
+              }
+            }
+
+            if (delCount < keyLimitPerTask) {
+              // Deep clean is completed, we can update the SnapInfo.
+              deepCleanedSnapshots.add(currSnapInfo.getTableKey());
+            }
+
+            if (!keysToPurge.isEmpty()) {
+              processKeyDeletes(keysToPurge, currOmSnapshot.getKeyManager(),
+                  keysToModify, currSnapInfo.getTableKey());
+            }
+
+          }
+        }
+      }
+      updateDeepCleanedSnapshots(deepCleanedSnapshots);
+    }
+
+    private void updateDeepCleanedSnapshots(List<String> deepCleanedSnapshots) {
+      if (!deepCleanedSnapshots.isEmpty()) {
+        SnapshotPurgeRequest snapshotPurgeRequest = SnapshotPurgeRequest
+            .newBuilder()
+            .addAllUpdatedSnapshotDBKey(deepCleanedSnapshots)
+            .build();
+
+        OMRequest omRequest = OMRequest.newBuilder()
+            .setCmdType(Type.SnapshotPurge)
+            .setSnapshotPurgeRequest(snapshotPurgeRequest)
+            .setClientId(clientId.toString())
+            .build();
+
+        submitRequest(omRequest);
+      }
+    }
+
+    public void submitRequest(OMRequest omRequest) {
+      try {
+        if (isRatisEnabled()) {
+          OzoneManagerRatisServer server = getOzoneManager().getOmRatisServer();
+
+          RaftClientRequest raftClientRequest = RaftClientRequest.newBuilder()
+              .setClientId(clientId)
+              .setServerId(server.getRaftPeerId())
+              .setGroupId(server.getRaftGroupId())
+              .setCallId(getRunCount().get())
+              .setMessage(Message.valueOf(
+                  OMRatisHelper.convertRequestToByteString(omRequest)))
+              .setType(RaftClientRequest.writeRequestType())
+              .build();
+
+          server.submitRequest(omRequest, raftClientRequest);
+        } else {
+          getOzoneManager().getOmServerProtocol()
+              .submitRequest(null, omRequest);
+        }
+      } catch (ServiceException e) {
+        LOG.error("Snapshot Deleting request failed. " +
+            "Will retry at next run.", e);
+      }
+    }
```

Comment:
put this in `AbstractKeyDeletingService` to be shared with `SnapshotDeletingService`?

Reply:
I tried this, but strangely the request is not going to OM at all.

Reply:
hmm okay let's leave it for now.